src/app/users/users.component.ts

```diff
@@ -102,12 +101,15 @@ export class UsersComponent implements OnInit, AfterViewInit {
     });
   }
 
-  deleteRole(user: any, index: number) {
+  deleteRole(user: any, index: number, position: number) {
     // Make copy of user so UI doesn't change until DB change succeeds
-    const tempUser = { ...user, roles: [ ...user.roles ] };
+    let tempUser = user;
     tempUser.roles.splice(index, 1);
-    this.selectedRolesMap.set(tempUser.name, tempUser.roles);
     delete tempUser.selected;
+    if (tempUser.roles.length === 0) {
+      tempUser = { ...tempUser, oldRoles: [] };
+      this.allUsers.data.forEach(row => this.updateOldRole(row, position, tempUser));

```

Comment:
We should not update the `users` variable or the `this.allUsers` array until the `subscribe`, and we don't need to iterate over the entire array since the `users` variable still refers back to the object in that array.